core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,423 @@
+}
+
+/*
+ * software reset of the entire I2C module.
+ * The module is reset and disabled.
+ * Status register fields (IBSR) are cleared.
+
+ * @param[in] Base       Base Address of I2c controller's registers
+ */
+void i2c_reset(vaddr_t base)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+
+	i2c_io_or8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+	i2c_io_or8((vaddr_t)&regs->ibsr, (I2C_IBSR_IBAL | I2C_IBSR_IBIF));
+	i2c_io_and8((vaddr_t)&regs->ibcr, ~(I2C_IBCR_IBIE | I2C_IBCR_DMAEN));
+	i2c_io_and8((vaddr_t)&regs->ibic, (uint8_t)(~I2C_IBIC_BIIE));
+}
+
+static uint8_t i2c_get_ibfd(vaddr_t base, uint16_t clock_divisor)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+	uint8_t ibfd = 0; /* I2c Bus Frequency Divider Register */
+	const struct i2c_clock_divisor_pair *clock_divisor_pair = NULL;
+	uint32_t clock_dvisor_pair_size = 0;
+	uint32_t index = 0;
+
+	if (io_read8((vaddr_t)&regs->ibdbg) & I2C_IBDBG_GLFLT_EN) {
+		clock_divisor_pair = i2c_clock_divisor_glitch_enabled;
+		clock_dvisor_pair_size = ARRAY_SIZE
+			(i2c_clock_divisor_glitch_enabled);
+	} else {
+		clock_divisor_pair = i2c_clock_divisor_glitch_disabled;
+		clock_dvisor_pair_size = ARRAY_SIZE
+			(i2c_clock_divisor_glitch_disabled);
+	}
+
+	if (clock_divisor >
+	    clock_divisor_pair[clock_dvisor_pair_size - 1].divisor) {
+		ibfd = clock_divisor_pair[clock_dvisor_pair_size - 1].ibfd;
+	} else {
+		for (index = 0; index < clock_dvisor_pair_size; index++) {
+			if (clock_divisor_pair[index].divisor >=
+			    clock_divisor) {
+				ibfd = clock_divisor_pair[index].ibfd;
+				break;
+			}
+		}
+	}
+
+	return ibfd;
+}
+
+TEE_Result i2c_init(struct ls_i2c_data *i2c_data)
+{
+	struct i2c_regs *regs = NULL;
+	uint16_t clock_divisor = 0;
+	uint8_t ibfd = 0; /* I2c Bus Frequency Divider Register */
+	size_t size = 0;
+	int node = 0;
+	vaddr_t ctrl_base = 0;
+	void *fdt = NULL;
+
+	/*
+	 * First get the I2C Controller base address from the DTB
+	 * if DTB present and if the I2C Controller defined in it.
+	 */
+	fdt = get_embedded_dt();
+	if (!fdt) {
+		EMSG("Unable to get the Embedded DTB, I2C init failed\n");
+		return TEE_ERROR_GENERIC;
+	}
+
+	node = fdt_path_offset(fdt,
+			       i2c_controller_map[i2c_data->i2c_controller]);
+	if (node > 0) {
+		if (dt_map_dev(fdt, node, &ctrl_base, &size) < 0) {
+			EMSG("Unable to get virtual address");
+			return TEE_ERROR_GENERIC;
+		}
+	} else {
+		EMSG("Unable to get I2C offset node");
+		return TEE_ERROR_ITEM_NOT_FOUND;
+	}
+
+	i2c_data->base = ctrl_base;
+
+	regs = (struct i2c_regs *)ctrl_base;
+	if (I2C_ERRATUM_A009203_FIX)
+		i2c_erratum_a009203(ctrl_base);
+
+	clock_divisor = (i2c_data->i2c_bus_clock + i2c_data->speed - 1) /
+			i2c_data->speed;
+	ibfd = i2c_get_ibfd(ctrl_base, clock_divisor);
+
+	io_write8((vaddr_t)&regs->ibfd, ibfd);
+
+	i2c_reset(ctrl_base);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_bus_test_bus_busy(struct i2c_regs *regs, bool test_busy)
+{
+	uint32_t index = 0;
```

Comment:
Please use `unsigned int n` or `size_t n` instead, same for all the other places.